bin/clang/replacePragma.cc

```diff
@@ -183,19 +189,34 @@ SSTReplacePragma::activate(Decl *d)
 void
 SSTReplacePragma::activateFunctionDecl(FunctionDecl *d)
 {
-  if (d->hasBody()) activate(d->getBody());
+  if (d->hasBody()){
+    PushGuard<FunctionDecl*> pg(CompilerGlobals::astContextLists.enclosingFunctionDecls, d);
+    //we are replacing expressions in the function body
+    //do NOT set the body to null, many expressions may remain valid
+    activate(d->getBody());
+  }
 }
 
 void
 SSTReplacePragma::activateVarDecl(VarDecl *d)
 {
-  if (d->hasInit()) activate(d->getInit());
+  if (d->hasInit()){
+    try {
+      activate(d->getInit());
+    } catch (StmtDeleteException& e) {
+      d->setInit(nullptr);
+    }
+  }
 }
 
 void
 SSTReplacePragma::activateCXXRecordDecl(CXXRecordDecl *d)
 {
-  if (d->hasBody()) activate(d->getBody());
+  if (d->hasBody()){
+    auto* body = d->getBody();
+    if (body) activate(d->getBody());
```

Comment:
Could write as 
```cpp
if(auto* body = d->getBody()){
    activate(body);
}
```
This would be idiomatic for llvm and also avoid the second dereference.